### PR TITLE
695 - blog backend

### PR DESCRIPTION
**Pull request by omonk**
## Description - [Trello ticket](https://trello.com/c/sburipjF/695-create-blog-post-template)

- Creates export lambda for blog work
- Creates all methods required to export blog posts from medium
- Adds tests for all methods

### CODE REVIEW

File: src/functions/blogMethods/parse-html-to-markdown.js [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+/**
+ * When Gatsby builds, Mdx tries to render <React.Suspense/> as a component
+ * instead of just string. Here we just need to wrap in `` to get Mdx to ignore
+ * trying to render this component...
+ */
+turndownService.addRule('p', {
+  filter: 'p',
+  replacement: content => {
+    const newContent = content.replace(/<([^>]*)>/g, (_, match) => {
```
Comment (jeysal) [inline src/functions/blogMethods/parse-html-to-markdown.js]: Seems like standard HTML sanitizing for paragraph texts - use a library?

File: src/functions/blogMethods/publish-to-contentful.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
+/* eslint-disable no-console */
+const { default: Map } = require('apr-map')
+const Intercept = require('apr-intercept')
+
+const postToCmsKeysMap = {
+  content: 'md'
+}
+
+const locale = 'en-US'
```
Comment (jeysal) [inline src/functions/blogMethods/publish-to-contentful.js]: should be used in more occurrences throughout files

File: src/functions/blogMethods/publish-to-contentful.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
+/* eslint-disable no-console */
+const { default: Map } = require('apr-map')
+const Intercept = require('apr-intercept')
+
+const postToCmsKeysMap = {
```
Comment (jeysal) [inline src/functions/blogMethods/publish-to-contentful.js]: necessary indirection?

File: src/functions/blogMethods/publish-to-contentful.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    (acc, curr) => ({
+      ...acc,
+      [curr]: {
+        [locale]: post[postToCmsKeysMap[curr]] || post[curr]
+      }
+    }),
+    {}
+  )
+
+module.exports = async (posts, environment, allFields, postSlugIdMap) =>
```
Comment (jeysal) [inline src/functions/blogMethods/publish-to-contentful.js]: fail fast?

File: src/functions/blogMethods/transform-custom-mdx.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+const { default: Map } = require('apr-map')
+const Waterfall = require('apr-waterfall')
+
+const TransformImageData = require('./transform-image-data')
+const TransformIFrames = require('./transform-iframes')
+const TransformStrings = require('./transform-strings')
+const AddFrontMatter = require('./add-frontmatter.js')
+
+module.exports = async (posts, environment) => {
+  return Map(posts, async post =>
```
Comment (jeysal) [inline src/functions/blogMethods/transform-custom-mdx.js]: Some modules export with `Map`, some without

File: src/functions/blogMethods/transform-iframes.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
+const Got = require('got')
+
+const gistBuilder = id => `<Gist id="${id}" />`
```
Comment (jeysal) [inline src/functions/blogMethods/transform-iframes.js]: JSX w/ jsdom factory?

File: src/functions/blogMethods/transform-iframes.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  return occurrences
+}
+
+const getIframeContent = async url => {
+  const { url: forwardedUrl } = await Got(url)
+
+  switch (true) {
+    case forwardedUrl.includes('gist.github'): {
```
Comment (jeysal) [inline src/functions/blogMethods/transform-iframes.js]: `URL` api

File: src/functions/blogMethods/transform-image-data.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+  // Upload images to Contentful from medium url
+  const uploadedImageData = await Map(images, async image => {
+    let cmsUrl
+    let cmsCaption
+    let cmsAlt
+    let cmsFileName
+    let cmsAssetId
+
+    if (isProd) {
```
Comment (jeysal) [inline src/functions/blogMethods/transform-image-data.js]: Still needed?

File: src/functions/blogMethods/transform-strings.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+const toReplace = [
+  ['https://medium.com/yld-blog/', 'https://yld.io/blog/'],
+  ['https://medium.com/yld-blog', 'https://yld.io/blog/'],
+  ['https://medium.com/yld-engineering-blog/', 'https://yld.io/blog/'],
+  ['https://medium.com/yld-engineering-blog', 'https://yld.io/blog/']
+]
+
+module.exports = post => ({
+  ...post,
+  md: toReplace.reduce(
```
Comment (jeysal) [inline src/functions/blogMethods/transform-strings.js]: Can be one regex

File: tests/unit/lambdas/blogMethods/parse-xml-to-json.test.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
+const ParseXMLToJSON = require('../../../../src/functions/blogMethods/parse-xml-to-json')
+const data = require('./__mocks__/parse-xml-to-json')
```
Comment (jeysal) [inline tests/unit/lambdas/blogMethods/parse-xml-to-json.test.js]: `readFileSync`

File: src/functions/blogMethods/index.js [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  const space = await client.getSpace(CONTENTFUL_SPACE)
+  const environment = await space.getEnvironment(environmentName)
+  const contentType = await environment.getContentType('blogPost')
+
+  const { allFields, requiredFields } = getContentTypeFields(contentType)
+
+  const { items: cmsBlogPosts } = await environment.getEntries({
+    limit: 1000,
```
[the remark this answers is not in the document]
Comment (omonk) [reply src/functions/blogMethods/index.js]: Not if this system is in place for the next ~10 years. Looking at the docs I can't see how to get the number of entries of said content type...

File: src/functions/blogMethods/parse-html-to-markdown.js [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+ * which is then rendered to the page in their own way.
+ *
+ * Further along in the process we follow this link
+ * and get the final URL.
+ *
+ */
+turndownService.addRule('iframe', {
+  filter: 'iframe',
+  replacement: content => {
+    const [, href] = content.match(/href="(.*)">/)
```
[the remark this answers is not in the document]
Comment (omonk) [reply src/functions/blogMethods/parse-html-to-markdown.js]: yep, think this was written before the realisation that we had JSdom, much better to use attribute

File: src/functions/blogMethods/transform-image-data.js (hunk 8 above)
Comment (omonk) [reply src/functions/blogMethods/transform-image-data.js]: Necessary for testing locally

File: src/functions/blogMethods/index.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (omonk) [reply src/functions/blogMethods/index.js]:
```
"total": 148,
    "skip": 0,
    "limit": 1000```

File: src/functions/blogMethods/publish-to-contentful.js (hunk 3 above)
Comment (omonk) [reply src/functions/blogMethods/publish-to-contentful.js]: will change

File: src/functions/blogMethods/transform-custom-mdx.js (hunk 5 above)
Comment (omonk) [reply src/functions/blogMethods/transform-custom-mdx.js]: move index Map methods to blogMethods/

File: src/functions/blogMethods/index.js [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+const getContentTypeFields = ct =>
+  ct.fields.reduce(
+    ({ allFields = [], requiredFields = [] }, { required, id }) => ({
+      allFields: allFields.concat(id),
+      requiredFields: required ? requiredFields.concat(id) : requiredFields
+    }),
+    []
+  )
+
+module.exports = async data => {
```
Comment (omonk) [inline src/functions/blogMethods/index.js]: input single XML string

File: src/functions/blogMethods/transform-iframes.js (hunk 6 above)
Comment (jeysal) [reply src/functions/blogMethods/transform-iframes.js]: nah

File: src/functions/blogMethods/filter-posts-to-process.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+// We don't want to publish certain posts
+const restrictedPosts = [
+  // Content is too long for Contentful
+  'node-js-databases-using-redis-for-fun-and-profit',
+  'node-js-databases-using-couchdb'
+]
+
+const filterPostsToProcess = (posts, { cmsBlogPosts, requiredFields }) => {
+  const cmsPostSlugs = cmsBlogPosts.map(({ fields }) => fields.slug['en-US'])
```
Comment (jeysal) [inline src/functions/blogMethods/filter-posts-to-process.js]: extract `en-US` from all files into a constant somewhere

File: src/functions/blogMethods/index.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (jeysal) [reply src/functions/blogMethods/index.js]: added to trello card

File: src/functions/blogMethods/publish-to-contentful.js (hunk 2 above)
Comment (jeysal) [reply src/functions/blogMethods/publish-to-contentful.js]: other comment tracks this

**Comment by jeysal** [comment]
Just the `en-US` thing is open, then after a quick manual test I guess we can merge

**Pull request merged**